File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
            lowest_past = prices[0]
            max_profit = 0

            for price in prices:
        
                if (price - lowest_past) > max_profit:
                    max_profit = price - lowest_past
                
                if price < lowest_past:
                    lowest_past = price

            return max_profit

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {

        pair<int, int> buy = {prices[0], 0};

        int profit = 0;

        int size = prices.size();

        for (int i = 0; i < size; i++) {
            if (buy.first > prices[i]) {
                buy.first = prices[i];
                buy.second = i;
            }

            if (buy.first < prices[i]) {
                int newProfit = prices[i] - buy.first;

                if (newProfit > profit) {
                    profit = newProfit;
                }
            }
        }

        return profit;
    }
};